### PR TITLE
Bugfix to merging loop in PFEGAlgo

### PR DESCRIPTION
**Pull request by lgray**
Fixes a bug in the merging loop of PFEGAlgo where it was possible that EB and EE clusters end up in the same e/gamma candidate.

The rate of this bug appears to be one in a few billion.

### CODE REVIEW

**Comment by lgray** [comment]
@cmsbuild please test

File: RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc [hunk 1]
```diff
@@ -1445,6 +1445,14 @@ initializeProtoCands(std::list<PFEGammaAlgo::ProtoEGObject>& egobjs) {
 	 << "Found objects " << std::distance(mergestart,nomerge)
 	 << " to merge by links to the front!" << std::endl;
        for( auto roToMerge = mergestart; roToMerge != nomerge; ++roToMerge) {
+         //bugfix! L.Gray 14 Jan 2016 
+         // -- check that the front is still mergeable!
+         if( !mergeTest(*roToMerge) ) {
+            // back up one step and repartition(since we will increment later)
+           nomerge = std::partition(roToMerge--,ROs.end(),mergeTest);
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc]: is this one safe against running out of bounds?

Comment (slava77) [inline RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc]: If this condition is really rare and not really expected, please add a LogWarning here

Comment (lgray) [inline RecoParticleFlow/PFProducer/src/PFEGammaAlgo.cc]: Safe against running out of bounds, yes since it starts from begin()+1.
I'll add the log warning.

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by lgray** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/12939	 f6c612d
- the same code change as in #12940 (76X), the same topic branch
-  the bugfix is tested extensively in 76X; jenkins here shows no differences

**Comment by davidlange6** [comment]
+1

**Pull request merged**